api/handlers.go

```diff
@@ -31,3 +32,123 @@ func (api *API) userHandlerGET(w http.ResponseWriter, req *http.Request, _ httpr
 	}
 	WriteJSON(w, u)
 }
+
+// userUploadsHandler returns all uploads made by the current user.
+func (api *API) userUploadsHandler(w http.ResponseWriter, req *http.Request, _ httprouter.Params) {
+	sub, _, _, err := tokenFromContext(req)
+	if err != nil {
+		WriteError(w, err, http.StatusInternalServerError)
+		return
+	}
+	u, err := api.staticDB.UserBySub(req.Context(), sub)
+	// TODO Should we create the user in case it's not found?
+	if err != nil {
+		WriteError(w, err, http.StatusInternalServerError)
+		return
+	}
+	ups, err := api.staticDB.UploadsByUser(req.Context(), *u)
+	if err != nil {
+		WriteError(w, err, http.StatusInternalServerError)
+	}
+	WriteJSON(w, ups)
+}
+
+// userDownloadsHandler returns all downloads made by the current user.
+func (api *API) userDownloadsHandler(w http.ResponseWriter, req *http.Request, _ httprouter.Params) {
+	sub, _, _, err := tokenFromContext(req)
+	if err != nil {
+		WriteError(w, err, http.StatusInternalServerError)
+		return
+	}
+	u, err := api.staticDB.UserBySub(req.Context(), sub)
+	// TODO Should we create the user in case it's not found?
+	if err != nil {
+		WriteError(w, err, http.StatusInternalServerError)
+		return
+	}
+	ups, err := api.staticDB.DownloadsByUser(req.Context(), *u)
+	if err != nil {
+		WriteError(w, err, http.StatusInternalServerError)
+	}
+	WriteJSON(w, ups)
+}
+
+// trackUploadHandler registers a new upload in the system.
+func (api *API) trackUploadHandler(w http.ResponseWriter, req *http.Request, ps httprouter.Params) {
+	sl := ps.ByName("skylink")
+	sub, _, _, err := tokenFromContext(req)
```

Comment:
I don't really understand how we're doing the authentication of these routes. Is there a middleware of sorts that handles authentication. I'm not sure we need to be asserting that the user exists. If we authenticate a user we could set its id on the request scope and simply insert the "tracking document" here with that user id. Saving a trip to the database. I feel there's quite a lot of stuff going on here. Like 3 or 4 database calls? Is this being triggered asynchronously or is this a blocking action. On upload it doesn't matter much but on download I think it would be less than ideal to have this be a blocking action.

Reply:
1. We are validating the token in a middleware called `validate` that's defined in `routes.go`. A valid token does not mean that we have the user in the DB, only that Kratos knows about them.
2. Tracking documents and so on are not something I'm currently familiar with. That's a little unfortunate but I haven't had the time to read up on Mongo's functionality, so I'm using bare-bones approach for the moment. We can definitely improve on this in the future. For the moment delivering a functioning first version seems to be more important.
3. Auto-creating the user is something that's already done in the code. This is already touched upon in other discussions in this PR but I believe that's a good approach in our use case - if the user doesn't exist create a new free user and carry on the request.
4. This is blocking on the service's side. I'm not sure if it needs to be blocking on nginx level. It will happen both on upload and on download but it will happen only once (and after that we'll have the user in the DB).